Zend/zend_execute.c

```diff
@@ -4268,6 +4274,140 @@ static zend_never_inline int ZEND_FASTCALL zend_quick_check_constant(
 	return _zend_quick_get_constant(key, 0, 1 OPLINE_CC EXECUTE_DATA_CC);
 } /* }}} */
 
+static zend_always_inline uint32_t zend_get_arg_offset_by_name(
+		zend_function *fbc, zend_string *arg_name, void **cache_slot) {
+	if (EXPECTED(*cache_slot == fbc)) {
+		return *(uintptr_t *)(cache_slot + 1);
+	}
+
+	// TODO: Use a hash table?
```

Comment:
One thing to keep in mind for the hash table to work is that PHP doesn't perform any validation of argument names for internal functions.
That would interact badly with named params for end users - what are your thoughts on a separate RFC to assert that argument names are valid PHP identifiers when loading functions/methods?

So it's possible for a PECL extension to do the following - I've seen examples of this in published extensions

`public function get($arg, $arg, $arg)` with `$arg` repeated more than once, which would be impossible in a user-defined function or method. (`ZEND_ARG_INFO(0, "arg")` instead of `ZEND_ARG_INFO(0, arg)`)
`public function get($"arg")` where the arg contains symbols that would be invalid in a user-defined function or method
`public function get($0, $1)`

A hash table seems doable, as long as it's keeping the entries added first and has those sanity checks

Reply:
In master we do check that all arguments have names. Also checking that they are unique shouldn't be a problem.

The main issue with the hashtable is that I would prefer to compute it lazily, rather than store it for all functions, most of which will likely never be invoked via named args. But this is incompatible with opcache, so would have to use map ptr, which has some cost.

The linear scan is maybe okay in practice (especially as it is usually cached), but not really viable if we want to also accept parameter names from parent methods, as some have suggested, as looking those up would be very inefficient.

Reply:
I have added a check for duplicate parameter names now.